### PR TITLE
feat: Add OAuth2 authentication support

### LINKED ISSUE #94: feat: Support OAuth
Currently the tap only supports API tokens https://developer.atlassian.com/cloud/jira/platform/basic-auth-for-rest-apis/#get-an-api-token. We should add support for OAuth.

### PR DESCRIPTION
**Pull request by imprvhub**
This commit introduces support for OAuth2 authentication in the tap-jira Singer tap.

Key changes include:

- Added a new `JiraOAuth2Authenticator` class in `authenticators.py` to handle the OAuth2
  authentication flow. This class manages access tokens, refresh tokens, and token expiration.

- Updated the `JiraStream` class in `client.py` to use a new `JiraAuthenticator` factory class.
  This allows the tap to seamlessly switch between basic and OAuth2 authentication based on the
  provided configuration.

- Added new OAuth2 related configuration options to the `TapJira` class in `tap.py`, including
  `client_id`, `client_secret`, `access_token`, and `refresh_token`. The `config_jsonschema` was
  updated to reflect these new optional properties and the new `auth_type` property.

- Added new unit tests in `tests/test_authenticators.py` to verify the behavior of the new
  `JiraOAuth2Authenticator` class, including testing token refresh and error handling.

- Updated the existing unit tests in `tests/test_core.py` to accommodate the new authentication
  options and to ensure the tap's core functionality remains intact.

This addition provides users with a more secure and flexible authentication option when using
the tap-jira. OAuth2 authentication is widely used and recommended by the Jira API for
server-to-server integrations.

The new code follows the existing design patterns and best practices used throughout the tap.
Exception handling and error messages have been included to provide clear feedback in case of
authentication issues.

Closes #94

### CODE REVIEW

**Comment by pnadolny13** [comment]
@imprvhub thanks for the contribution! I added myself and Edgar as reviewers.

@ReubenFrankel I cant technically add you as a reviewer but I wanted to ping you on this since you were asking about it in https://github.com/MeltanoLabs/tap-jira/pull/86#issuecomment-2480865503. Feel free to leave comments and thoughts!

**Comment by ReubenFrankel** [comment]
Thanks for this. As Pat pointed out in #86 it feels a bit redundant to support `access_token` given that they are short-lived, but no harm in accepting it - especially now there is full OAuth support and could eventually work with [write-back](https://github.com/meltano/sdk/issues/106).

File: tap_jira/tap.py [hunk 1]
```diff
@@ -15,33 +16,70 @@ class TapJira(Tap):
 
     config_jsonschema = th.PropertiesList(
```
Comment (ReubenFrankel) [inline tap_jira/tap.py:17]: Do we want to represent the desired auth flow using `singer_sdk.typing.DiscriminatedUnion` as it was previously?

https://github.com/MeltanoLabs/tap-jira/pull/86/files#diff-9a8a5da31511510c569573541d7460963e518208e1f71162cf6847378a12dfb4L33-L48

**Comment by ReubenFrankel** [comment]
Did you mean to close this? @imprvhub

**Comment by imprvhub** [comment]
@ReubenFrankel  Yes, I intentionally closed this PR. After reviewing the complete interaction history from [#94](https://github.com/MeltanoLabs/tap-jira/issues/94) to here, I realized this might not be the right time or context for my contribution. While I appreciate the technical feedback provided, certain aspects of the collaboration process made me reconsider my participation in this project. I wish you success with the OAuth2 implementation.

**Pull request closed without merging**